File: challenges/elixir/.claude/archive/add_docs.py
#!/usr/bin/env python3
"""
Script to add @moduledoc and @doc to Elixir code blocks in markdown files
"""

import os
import re
import sys
from pathlib import Path

class DocAdder:
    def __init__(self, root_dir):
        self.root_dir = Path(root_dir)
        self.stats = {
            'files_processed': 0,
            'moduledocs_added': 0,
            'docs_added': 0
        }

    def run(self):
        """Process all markdown files"""
        md_files = list(self.root_dir.rglob('*.md'))

        for md_file in md_files:
            # Skip dependency files
            if 'deps' in str(md_file):
                continue

            self.process_file(md_file)

        self.print_summary()
        self.save_log()

    def process_file(self, file_path):
        """Process a single markdown file"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
        except Exception as e:
            print(f"Error reading {file_path}: {e}")
            return

        new_content, moduledocs, docs = self.add_docs_to_content(content)

        if moduledocs > 0 or docs > 0:
            try:
                with open(file_path, 'w', encoding='utf-8') as f:
                    f.write(new_content)
                print(f"✓ {file_path}: +{moduledocs} @moduledoc, +{docs} @doc")
                self.stats['files_processed'] += 1
                self.stats['moduledocs_added'] += moduledocs
                self.stats['docs_added'] += docs
            except Exception as e:
                print(f"Error writing {file_path}: {e}")

    def add_docs_to_content(self, content):
        """Add @moduledoc and @doc to code blocks"""
        lines = content.split('\n')
        new_lines = []
        in_code_block = False
        moduledocs_count = 0
        docs_count = 0
        i = 0

        while i < len(lines):
            line = lines[i]

            # Check for elixir code block start
            if line.strip().startswith('```elixir'):
                in_code_block = True
                new_lines.append(line)
                i += 1
                continue

            # Check for code block end
            if line.strip().startswith('```') and in_code_block:
                in_code_block = False
                new_lines.append(line)
                i += 1
                continue

            # Process lines inside code blocks
            if in_code_block:
                stripped = line.lstrip()
                indent = line[:len(line) - len(stripped)]

                # Handle defmodule - check if next line is @moduledoc
                if re.match(r'^defmodule\s+\w+', stripped):
                    # Look ahead to see if @moduledoc exists
                    has_moduledoc = False
                    if i + 1 < len(lines):
                        next_line = lines[i + 1].lstrip()
                        if next_line.startswith('@moduledoc') or next_line.startswith('@doc'):
                            has_moduledoc = True

                    new_lines.append(line)

                    if not has_moduledoc:
                        # Extract module name for better doc
                        match = re.match(r'^defmodule\s+(\w+)', stripped)
                        module_name = match.group(1) if match else 'Module'
                        formatted_name = self.format_name(module_name)
                        moduledoc_line = f'{indent}@moduledoc "{formatted_name}"'
                        new_lines.append(moduledoc_line)
                        moduledocs_count += 1

                    i += 1
                    continue

                # Handle def - check if it's a public function
                if re.match(r'^def\s+\w+', stripped) and not re.match(r'^def\s+(handle_call|handle_cast|init|handle_info|code_change|terminate)', stripped):
                    # Check if previous line is @doc
                    has_doc = False
                    if new_lines and new_lines[-1].lstrip().startswith('@doc'):
                        has_doc = True

                    if not has_doc:
                        # Extract function name
                        match = re.match(r'^def\s+(\w+)', stripped)
                        func_name = match.group(1) if match else 'function'
                        formatted_name = self.format_name(func_name)
                        doc_line = f'{indent}@doc "{formatted_name}"'
                        new_lines.append(doc_line)
                        docs_count += 1

                    new_lines.append(line)
                    i += 1
                    continue

            new_lines.append(line)
            i += 1

        return '\n'.join(new_lines), moduledocs_count, docs_count

    def format_name(self, name):
        """Convert camelCase to readable name"""
        # Insert space before uppercase letters
        spaced = re.sub(r'(?<!^)(?=[A-Z])', ' ', name)
        return spaced.lower()

    def print_summary(self):
        """Print processing summary"""
        stats = self.stats
        print("\n" + "=" * 60)
        print(f"Code Quality (Docs): {stats['files_processed']} archivos procesados")
        print(f"{stats['moduledocs_added']} @moduledoc agregados")
        print(f"{stats['docs_added']} @doc agregados")
        print("=" * 60 + "\n")

    def save_log(self):
        """Save processing log"""
        log_dir = self.root_dir / '.claude'
        log_dir.mkdir(exist_ok=True)
        log_file = log_dir / 'code-quality-docs.log'

        import datetime
        timestamp = datetime.datetime.utcnow().isoformat()

        log_content = f"""Code Quality Documentation Automation Log
========================================
Timestamp: {timestamp}

Summary:
- Files processed: {self.stats['files_processed']}
- @moduledoc added: {self.stats['moduledocs_added']}
- @doc added: {self.stats['docs_added']}

Status: COMPLETED
"""

        with open(log_file, 'w', encoding='utf-8') as f:
            f.write(log_content)

        print(f"Log saved to: {log_file}")

if __name__ == '__main__':
    root = "/Users/consulting/Documents/consulting/infra/challenges/elixir"
    adder = DocAdder(root)
    adder.run()

File: challenges/elixir/.claude/archive/add_docs.exs
#!/usr/bin/env elixir
# Script para agregar @moduledoc y @doc automáticamente a archivos .md

defmodule DocAdder do
  @moduledoc "Automates adding @moduledoc and @doc to Elixir code blocks in markdown files"

  def run(root_dir) do
    md_files = find_md_files(root_dir)

    stats = Enum.reduce(md_files, %{files: 0, moduledocs: 0, docs: 0}, fn file, acc ->
      process_file(file, acc)
    end)

    print_summary(stats)
    save_log(stats, root_dir)
    stats
  end

  defp find_md_files(root_dir) do
    root_dir
    |> Path.expand()
    |> Stream.unfold(&next_file/1)
    |> Enum.filter(&String.ends_with?(&1, ".md"))
  end

  defp next_file(dir) do
    case File.ls(dir) do
      {:ok, files} ->
        {dir, files}
        |> expand_paths()
        |> Enum.split_with(&File.dir?/1)
        |> case do
          {dirs, files} ->
            files_with_path = Enum.map(files, &Path.join(dir, &1))
            remaining = dirs ++ []
            {files_with_path, remaining}
        end
      {:error, _} ->
        nil
    end
  end

  defp expand_paths({dir, files}) do
    Enum.map(files, &Path.join(dir, &1))
  end

  defp process_file(file, acc) do
    case File.read(file) do
      {:ok, content} ->
        {new_content, moduledocs, docs} = add_docs_to_content(content)

        if moduledocs > 0 or docs > 0 do
          File.write(file, new_content)
          IO.puts("✓ #{file}: +#{moduledocs} @moduledoc, +#{docs} @doc")
          %{acc | files: acc.files + 1, moduledocs: acc.moduledocs + moduledocs, docs: acc.docs + docs}
        else
          acc
        end

      {:error, _} ->
        acc
    end
  end

  defp add_docs_to_content(content) do
    lines = String.split(content, "\n")
    {new_lines, moduledocs, docs} = process_lines(lines, [], false, 0, 0)
    {Enum.join(new_lines, "\n"), moduledocs, docs}
  end

  defp process_lines([], acc, _in_code, moduledocs, docs) do
    {Enum.reverse(acc), moduledocs, docs}
  end

  defp process_lines([line | rest], acc, in_code, moduledocs, docs) do
    cond do
      # Detectar inicio de bloque de código Elixir
      String.starts_with?(line, "```elixir") ->
        process_lines(rest, [line | acc], true, moduledocs, docs)

      # Detectar fin de bloque de código
      String.starts_with?(line, "```") and in_code ->
        process_lines(rest, [line | acc], false, moduledocs, docs)

      # Procesar líneas dentro de bloques de código
      in_code ->
        trimmed = String.trim_start(line)
        cond do
          # Detectar defmodule sin @moduledoc
          String.starts_with?(trimmed, "defmodule ") ->
            # Buscar si la siguiente línea no es @moduledoc
            case rest do
              [next | _] ->
                next_trimmed = String.trim_start(next)
                if String.starts_with?(next_trimmed, "@moduledoc") or String.starts_with?(next_trimmed, "@doc") do
                  process_lines(rest, [line | acc], in_code, moduledocs, docs)
                else
                  # Agregar @moduledoc
                  indent = get_indent(line)
                  module_name = extract_module_name(line)
                  moduledoc_line = "#{indent}@moduledoc \"Implements #{format_name(module_name)}\""
                  process_lines(rest, [moduledoc_line, line | acc], in_code, moduledocs + 1, docs)
                end
              [] ->
                process_lines(rest, [line | acc], in_code, moduledocs, docs)
            end

          # Detectar def sin @doc (pero no handle_call, handle_cast, init)
          String.match?(trimmed, ~r/^def\s+\w+/) ->
            callback? = String.match?(trimmed, ~r/^def\s+(handle_call|handle_cast|init|handle_info|code_change|terminate)/)
            if callback? do
              process_lines(rest, [line | acc], in_code, moduledocs, docs)
            else
              case acc do
                ["@doc" <> _ | _] ->
                  # Ya tiene @doc
                  process_lines(rest, [line | acc], in_code, moduledocs, docs)
                _ ->
                  # Agregar @doc
                  indent = get_indent(line)
                  func_name = extract_function_name(line)
                  doc_line = "#{indent}@doc \"#{format_name(func_name)}\""
                  process_lines(rest, [line, doc_line | acc], in_code, moduledocs, docs + 1)
              end
            end

          true ->
            process_lines(rest, [line | acc], in_code, moduledocs, docs)
        end

      true ->
        process_lines(rest, [line | acc], in_code, moduledocs, docs)
    end
  end

  defp get_indent(line) do
    case Regex.scan(~r/^(\s*)/, line) do
      [[indent, _]] -> indent
      _ -> ""
    end
  end

  defp extract_module_name(line) do
    case Regex.run(~r/defmodule\s+(\w+)/, line) do
      [_, name] -> name
      _ -> "Module"
    end
  end

  defp extract_function_name(line) do
    case Regex.run(~r/def\s+(\w+)/, line) do
      [_, name] -> name
      _ -> "function"
    end
  end

  defp format_name(name) do
    name
    |> String.split(~r/(?=[A-Z])/)
    |> Enum.join(" ")
    |> String.downcase()
    |> String.capitalize()
  end

  defp print_summary(%{files: files, moduledocs: moduledocs, docs: docs}) do
    IO.puts("\n" <> String.duplicate("=", 60))
    IO.puts("Code Quality (Docs): #{files} archivos procesados")
    IO.puts("#{moduledocs} @moduledoc agregados")
    IO.puts("#{docs} @doc agregados")
    IO.puts(String.duplicate("=", 60) <> "\n")
  end

  defp save_log(%{files: files, moduledocs: moduledocs, docs: docs}, root_dir) do
    log_file = Path.join(root_dir, ".claude/code-quality-docs.log")
    timestamp = DateTime.utc_now() |> DateTime.to_string()

    log_content = """
    Code Quality Documentation Automation Log
    ========================================
    Timestamp: #{timestamp}

    Summary:
    - Files processed: #{files}
    - @moduledoc added: #{moduledocs}
    - @doc added: #{docs}

    Status: COMPLETED
    """

    File.mkdir_p(Path.dirname(log_file))
    File.write(log_file, log_content)
    IO.puts("Log saved to: #{log_file}")
  end
end

# Ejecutar
root = "/Users/consulting/Documents/consulting/infra/challenges/elixir"
DocAdder.run(root)
